man/lxpanelctl.1
'\" t
.\"     Title: LXPANEL
.\"    Author: Ying-Chun Liu
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: March  2, 2008
.\"    Manual: http://LXDE.org
.\"    Source: http://LXDE.org
.\"  Language: English
.\"
.TH "LXPANEL" "1" "March, 2016" "http://LXDE\&.org" "http://LXDE.org"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
lxpanelctl \- controller for lxpanel\&.
.SH "SYNOPSIS"
.HP \w'\fBlxpanelctl\fR\ 'u
\fBlxpanelctl\fR {command}
.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBlxpanelctl\fR
command\&.
.PP
\fBlxpanelctl\fR
is a program that controls lxpanel\&.
.SH "COMMANDS"
.PP
\fBmenu\fR
.RS 4
Show the system menu\&.
.RE
.PP
\fBshow\fR
.RS 4
Show the panel in autohide mode\&.
.RE
.PP
\fBrun\fR
.RS 4
Show the run dialog\&.
.RE
.PP
\fBconfig\fR
.RS 4
Show the panel configuration dialog\&. If more than one panel is configured, it is undefined which panel configuration dialog will be shown\&.
.RE
.PP
\fBrestart\fR
.RS 4
Restart lxpanel\&.
.RE
.PP
\fBexit\fR
.RS 4
Exit lxpanel\&.
.RE
.PP
\fBcommand \fR[\fB\-\-panel=\fR[\fIN\fR:]\fIedge\fR] \fIplugin\fR \fIcommand\fR
.RS 4
Send a \fIcommand\fR to a \fIplugin\fR\&. Optionally monitor number \fIN\fR
(in range 1 to 8) and \fIedge\fR (left, right, top or bottom) can be set,
otherwise \fIcommand\fR will be send to first \fIplugin\fR found in any panel\&.
.RE
.SH "SEE ALSO"
.PP
lxpanel (1)\&.
.SH "AUTHOR"
.PP
This manual page was written by paulliu
grandpaul@gmail\&.com
for the
Debian
system (but may be used by others)\&. Permission is granted to copy, distribute and/or modify this document under the terms of the
GNU
General Public License, Version 2 any later version published by the Free Software Foundation\&.
.PP
On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common\-licenses/GPL-2\&.
.PP
Updated by Andriy Grytsenko (LStranger)\&.
.SH "AUTHOR"
.PP
\fBYing\-Chun Liu\fR
.RS 4
Author.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2008 paulliu
.br
Copyright \(co 2016 LStranger
.br


ChangeLog
0.11.1
-------------------------------------------------------------------------
* Less restrictive location query and more detailed display of results
    in the weather plugin.
* Minor fixes related to the new icons.
* Updated translations.

0.11.0
-------------------------------------------------------------------------
* Added tooltips to application launch bar and panel settings.
* Added option to battery monitor to hide number of battery in tooltip.
* Added option to resource monitors to show cached and buffered memory as
    used.
* Improved translations and design fix of ALSA volume control popup.
* Made netstatus_icon_set_tooltips_enabled() work.
* Fixed a sometimes incorrectly placed icon in the netstatus plugin.
* Improved weather plugin: correct local time, normal closing of dialog,
    easier to recognize information in dialog and tooltip, better quality
    of weather conditions icon and image.
* HiDPI support on GTK+ 3.
* Extended battery life check to about a minute before triggering battery
    alarm.
* Added an option to disable square brackets around minimized window names.
* Reworked xkb plugin.
* Improved tracking of minimized windows.
* Implemented check for sensors returning error instead of temperature.
* Restored disabled CPUFreq plugin.
* Improved icons and updated icon names.
* Updated the default configuration files.
* Various fixes for GTK+ 3 compatibility.
* Various bug fixes and improvements.
* Updated translations.

0.10.1
-------------------------------------------------------------------------
* Fixed battery alarm when measurement of current is missing.
* Fixed spelling errors on "allow to" in plugins descriptions, and "GTK2+" to
    more correct "GTK+".
* Fixed battery indication when charge_full and energy_full are not available.
* Corrected scaling of volume plugin for HiDPI monitors.
* Fixed connection name dropdown list in netstatus plugin.

0.10.0
-------------------------------------------------------------------------
* Fixed delayed application of font cnahges on dclock and weather plugins.
* Allowed switching workplaces by scrolling over the pager.
* Added new API: lxpanel_draw_label_text_with_color().
* Fixed not-applicable font setting on thermal plugin.
* Fixed crash on right-click task button after some task was killed using menu
    on another task button.
* Fixed crash after color was removed from monitor plugin configuration.
* Aligned used memory calculation in lxpanel monitor with 'free' command line
    utility.
* Added possibility to use few weather providers for weather widget.
* Disabled Yahoo! Weather because they removed free API support.
* Added OpenWeatherMap weather provider.
* Fixed wrong application position on taskbar in multihead configuration.
* Fixed battery color (must be green) when battery is fully charged.
* Fixed crash at task plugin when all windows lose focus.
* Disambiguated keyboard layout names to follow variants.

0.9.3
-------------------------------------------------------------------------
* Finally fixed battery calculations, it missed Power so left time was
    always at zero.
* Fixed task window raising on drop file on it (was broken since 0.9.0).

0.9.2
-------------------------------------------------------------------------
* Fixed battery selection if battery detached but another is available.
* Fixed battery percentage calculations if charge isn't available but
    energy is, it appears broken in 0.9.0.
* Fixed wrong state change on task button when hovered, it shouldn't
    receive selected state until window has urgency.
* Fixed wrong flat button appearance when hovered, it wasn't flat.

0.9.1
-------------------------------------------------------------------------
* Fixed build failure without ALSA.
* Added alternative mixers if build using OSS.
* Fixed panel flicker with some window managers.

0.9.0
-------------------------------------------------------------------------
* Added support for laptops with multiple batteries.
* Added a way to scroll through workspaces by using the scroll wheel over
    the deskno plugin.
* Bunch of fixes for GTK+ 3.0 compatibility.
* Fixed outdated lxpanel man page.
* Fixed weather locations query (since YQL was changed recently).
* Converted space plugin into internal PanelSpace widget in liblxpanel.
* Converted task button into widget class to avoid ambiguous code.
* Added support for third-party plugins localized descriptions.
* Implemented "lxpanelctl command ..." to send message to panel plugin.
* Added APIs for PanelIconGrid to draw focus on drop.
* Converted launch button into widget class for more easy management.
* Adjusted PanelIconGrid: min spacing is now 1, no_window flag is now
    left unset by default, spacing is taken into account on overall size
    to be consistent on chosen orientation.
* Implemented dropping file or folder onto launch bar to create launcher.
* Implemented dragging launchers around launchbars (between panels too).
* Implemented dropping file or folder onto free panel space to create
    launcher.
* Implemented dragging launchers from launchbar to some free panel space.
    When launcher dropped onto another place and former place's launchbar
    becomes empty, it will not be replaced with a bootstrapping icon (as
    it happens when last launcher is removed via configuration dialog)
    but launchbar will be removed from the panel instead, that behavior
    is more obvious and sensible.
* Added stripping scheme from URI menu://applications/... when a launcher
    is created, only menu-id itself should be used.
* Implemented moving plugins across panel by means of middle button drag
    and drop, i.e. middle-click it then drag, and plugin will follow the
    mouse until middle button is released.
* Merged 'volume' and 'volumealsa' plugins into single plugin which uses
    the same interface but different driver, in dependency on how it was
    built - either OSS or ALSA interface is used. Name 'volume' is used

src/lxpanelctl.c
/*
 * Copyright (C) 2006 Hong Jen Yee (PCMan) <pcman.tw@gmail.com>
 *               2006 Jim Huang <jserv.tw@gmail.com>
 *               2009-2010 Marty Jack <martyj19@comcast.net>
 *               2016 Andriy Grytsenko <andrej@rep.kiev.ua>
 *
 * This file is a part of LXPanel project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/* Remote controller of lxpanel */

#include "lxpanelctl.h"
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

static Display* dpy;

static const char usage[] =
        "\nlxpanelctl - LXPanel Controller\n"
        "Usage: lxpanelctl <command>\n\n"
        "Available commands:\n"
        "menu\t\t\tshow system menu\n"
        "show\t\t\tshow panel if hidden\n"
        "run\t\t\tshow run dialog\n"
        "config\t\t\tshow configuration dialog\n"
        "restart\t\t\trestart lxpanel\n"
        "exit\t\t\texit lxpanel\n"
        "command <plugin> <cmd>\tsend a command to a plugin\n\n";

static int get_cmd( const char* cmd )
{
    if( ! strcmp( cmd, "menu") )
        return LXPANEL_CMD_SYS_MENU;
    if( ! strcmp( cmd, "show") )
        return LXPANEL_CMD_SHOW_PANEL;
    else if( ! strcmp( cmd, "run") )
        return LXPANEL_CMD_RUN;
    else if( ! strcmp( cmd, "config") )
        return LXPANEL_CMD_CONFIG;
    else if( ! strcmp( cmd, "restart") )
        return LXPANEL_CMD_RESTART;
    else if( ! strcmp( cmd, "exit") )
        return LXPANEL_CMD_EXIT;
    else if( ! strcmp( cmd, "command") )
        return LXPANEL_CMD_COMMAND;
    return -1;
}

/* format: either "<edge>" or "<num>:<edge>" */
static int parse_id(const char *arg, int *mon)
{
    char *end;
    long lmon = strtoul(arg, &end, 10);

    if (*end == ':')
        arg = (const char *)end + 1;
    else
        lmon = 0;
    *mon = lmon;
    if (strcmp(arg, "top") == 0)
        return EDGE_TOP;
    if (strcmp(arg, "bottom") == 0)
        return EDGE_BOTTOM;
    if (strcmp(arg, "left") == 0)
        return EDGE_LEFT;
    if (strcmp(arg, "right") == 0)
        return EDGE_RIGHT;
    return EDGE_NONE;
}

int main( int argc, char** argv )
{
    char *display_name = (char *)getenv("DISPLAY");
    XEvent ev;
    Window root;
    Atom cmd_atom;
    int cmd;
    /* int restart; */
    /* target of message, it's XClientMessageEvent::b[1]
     * valid only if XClientMessageEvent::b[0] == LXPANEL_CMD_COMMAND */
    uint8_t target;

    if( argc < 2 )
    {
        printf( usage );
        return 1;
    }

    /*
    if( restart = !strcmp( argv[1], "restart" ) )
        argv[1] = "exit";
    */

    if( ( cmd = get_cmd( argv[1] ) ) == -1 )
        return 1;

    if (cmd == LXPANEL_CMD_COMMAND && argc < 4)
    {
        printf( usage );
        return 1;
    }

    dpy = XOpenDisplay(display_name);
    if (dpy == NULL) {
        printf("Cant connect to display: %s\n", display_name);
        exit(1);
    }
    root = DefaultRootWindow(dpy);
    cmd_atom = XInternAtom(dpy, "_LXPANEL_CMD", False);
    memset(&ev, '\0', sizeof ev);
    ev.xclient.type = ClientMessage;
    ev.xclient.window = root;
    ev.xclient.message_type = cmd_atom;
    ev.xclient.format = 8;

    ev.xclient.data.b[0] = cmd;

    if (cmd == LXPANEL_CMD_COMMAND)
    {
        int i = 2;
        if (argc > 4 && strncmp(argv[2], "--panel=", 8) == 0)
        {
            int monitor;
            int edge = parse_id(argv[2] + 8, &monitor);
            /* edge: EDGE_NONE ..., monitor: 0 - none, 1...8 - selected */
            target = ((edge & 0x7) << 4) + (monitor & 0xf);
            i++;
        }
        else
            target = (EDGE_NONE << 4) + 0; /* edge: none, monitor: none */
        ev.xclient.data.b[1] = target;
        snprintf(&ev.xclient.data.b[2], 18, "%s\t%s", argv[i], argv[i+1]);
    }

    XSendEvent(dpy, root, False,
               SubstructureRedirectMask|SubstructureNotifyMask, &ev);
    XSync(dpy, False);
    XCloseDisplay(dpy);

/*
    if( restart ) {
        system( PACKAGE_BIN_DIR "/lxpanel &" );
    }
*/
    return 0;
}

